app/services/PaypalService.scala

```diff
@@ -0,0 +1,157 @@
+package services
+
+import java.util.UUID
+
+import com.gu.i18n.CountryGroup
+import com.paypal.api.payments._
+import com.paypal.base.Constants
+import com.paypal.base.rest.{APIContext, PayPalRESTException}
+
+import scala.collection.JavaConverters._
+import com.typesafe.config.Config
+import data.ContributionData
+import models.{ContributionMetaData, Contributor}
+import org.joda.time.DateTime
+import play.api.Logger
+import views.support.ChosenVariants
+
+import scala.util.Try
+
+case class PaypalCredentials(clientId: String, clientSecret: String)
+
+case class PaypalApiConfig(
+  envName: String,
+  paypalMode: String,
+  baseReturnUrl: String,
+  credentials: PaypalCredentials,
+  paypalWebhookId: String
+)
+
+object PaypalApiConfig {
+  def from(config: Config, environmentName: String, variant: String = "api") = PaypalApiConfig(
+    envName = environmentName,
+    credentials = PaypalCredentials(config.getString("clientId"), config.getString("clientSecret")),
+    paypalMode = config.getString("paypalMode"),
+    baseReturnUrl = config.getString("baseReturnUrl"),
+    paypalWebhookId = config.getString("paypalWebhookId")
+  )
+}
+
+class PaypalService(config: PaypalApiConfig, contributionData: ContributionData) {
+  val description = "Contribution to the guardian"
+  val credentials = config.credentials
+
+
+  def apiContext: APIContext = new APIContext(credentials.clientId, credentials.clientSecret, config.paypalMode)
+
+  def getAuthUrl(
+    amount: BigDecimal,
+    countryGroup: CountryGroup,
+    contributionId: String,
+    cmp: Option[String],
+    intCmp: Option[String],
+    ophanId: Option[String]
+  ): Either[String, String] = {
+
+    def returnUrl: String = {
+      val extraParams = List(
+        cmp.map(value => s"CMP=$value"),
+        intCmp.map(value => s"INTCMP=$value"),
+        ophanId.map(value => s"ophanId=$value")
+      ).flatten match {
+        case Nil => ""
+        case params => params.mkString("?", "&", "")
+      }
+      s"${config.baseReturnUrl}/paypal/${countryGroup.id}/execute$extraParams"
+    }
+
+    val cancelUrl = config.baseReturnUrl
+    val currencyCode = countryGroup.currency.toString
+    val paypalAmount = new Amount().setCurrency(currencyCode).setTotal(amount.toString)
+    val item = new Item().setDescription(description).setCurrency(currencyCode).setPrice(amount.toString).setQuantity("1")
+    val itemList = new ItemList().setItems(List(item).asJava)
+    val transaction = new Transaction
+    transaction.setAmount(paypalAmount)
+    transaction.setDescription(description)
+    transaction.setCustom(contributionId)
+    transaction.setItemList(itemList)
+
+    val transactions = List(transaction).asJava
+
+    val payer = new Payer().setPaymentMethod("paypal")
+    val redirectUrls = new RedirectUrls().setCancelUrl(cancelUrl).setReturnUrl(returnUrl)
+
+    val payment = new Payment().setIntent("sale").setPayer(payer).setTransactions(transactions).setRedirectUrls(redirectUrls)
+    try {
```

Comment:
Using the Try class here will allow the compile to type check the code rather than the key word try.  

[Error handling with Try](http://danielwestheide.com/blog/2012/12/26/the-neophytes-guide-to-scala-part-6-error-handling-with-try.html)